src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonPatternDetector.cs

```diff
@@ -0,0 +1,307 @@
+               syntaxFacts.IsArgument(token.Parent.Parent) &&
+               syntaxFacts.IsInvocationExpression(token.Parent.Parent.Parent.Parent);
+
+        public bool IsDefinitelyJson(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotJson(token, _syntaxFacts))
+            {
+                return false;
+            }
+
+            if (HasJsonLanguageComment(token, _syntaxFacts, out _))
+            {
+                return true;
+            }
+
+            if (!IsMethodArgument(token, _syntaxFacts))
+            {
+                return false;
+            }
+
+            var stringLiteral = token;
+            var literalNode = stringLiteral.Parent;
+            var argumentNode = literalNode.Parent;
+            Debug.Assert(_syntaxFacts.IsArgument(argumentNode));
+
+            var argumentList = argumentNode.Parent;
+            var invocationOrCreation = argumentList.Parent;
+            if (_syntaxFacts.IsInvocationExpression(invocationOrCreation))
+            {
+                var invokedExpression = _syntaxFacts.GetExpressionOfInvocationExpression(invocationOrCreation);
+                var name = GetNameOfInvokedExpression(invokedExpression);
+                if (_syntaxFacts.StringComparer.Equals(name, _methodNameOfInterest))
+                {
+                    // Is a string argument to a method that looks like it could be a Regex method.  
+                    // Need to do deeper analysis
+                    var method = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken).GetAnySymbol();
+                    if (method != null &&
+                        method.DeclaredAccessibility == Accessibility.Public &&
+                        method.IsStatic &&
+                        _typesOfInterest.Contains(method.ContainingType))
+                    {
+                        return AnalyzeStringLiteral(
+                            stringLiteral, argumentNode, cancellationToken);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public JsonTree TryParseJson(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotJson(token, _syntaxFacts))
+            {
+                return null;
+            }
+
+            HasJsonLanguageComment(token, _syntaxFacts, out var strict);
+
+            var chars = _virtualCharService.TryConvertToVirtualChars(token);
+            if (chars.IsDefaultOrEmpty)
+            {
+                return null;
+            }
+
+            return JsonParser.TryParse(chars, strict);
+        }
+
+        private bool AnalyzeStringLiteral(
+            SyntaxToken stringLiteral, SyntaxNode argumentNode,
+            CancellationToken cancellationToken)
+        {
+            var parameter = _semanticFacts.FindParameterForArgument(_semanticModel, argumentNode, cancellationToken);
```

Comment:
It looks like `_semanticFacts` is only used in this method. Consider passing in as a parameter.

Reply:
I prefer this approach.  All the data that stays constant for the Document is held as fields of the class, and the only data that is passed in is the stuff that can actually vary from call to call.

Reply:
#bydesign